File: neo/backend/src/app/api/v1/events.py
"""API routes for events."""

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.repositories import workspace_repository
from app.repositories.event_repository import EventRepository
from app.schemas.event import (
    EventCreate,
    EventListResponse,
    EventResponse,
    EventUpdate,
)
from app.schemas.response import ApiResponse

router = APIRouter(prefix="/workspaces/{workspace_code}/events", tags=["events"])


def _get_workspace_id(
    workspace_code: str,
    db: Session,
) -> int:
    """Get workspace ID by code, raise 404 if not found."""
    workspace = workspace_repository.get_workspace_by_code(db, workspace_code)

    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    # TODO: Check workspace membership
    return int(workspace.id)


def _to_event_response(event) -> EventResponse:
    """Convert Event model to EventResponse, excluding relationships."""
    return EventResponse(
        id=event.id,
        name=event.name,
        entity_name=event.entity_name,
        target_entity_name=event.target_entity_name,
        actor=event.actor,
        timestamp=event.timestamp,
        page_url=event.page_url,
        session_id=event.session_id,
        metadata=event.event_metadata,
        workspace_id=event.workspace_id,
        embedded_site_id=event.embedded_site_id,
        created_by=event.created_by,
        created_at=event.created_at,
        updated_at=event.updated_at,
    )


@router.get("", response_model=ApiResponse[EventListResponse])
def list_events(
    workspace_code: str,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    name: Optional[str] = Query(None, description="Event name search (partial match)"),
    entity_name: Optional[str] = Query(None, description="Entity name (exact match)"),
    actor: Optional[str] = Query(None, description="Actor search (partial match)"),
    timestamp_start: Optional[datetime] = Query(None, description="Start time (ISO 8601)"),
    timestamp_end: Optional[datetime] = Query(None, description="End time (ISO 8601)"),
    embedded_site_id: Optional[int] = Query(None, description="Filter by embedded site ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EventListResponse:
    """List events for a workspace with pagination and filtering."""
    workspace_id = _get_workspace_id(workspace_code, db)
    repo = EventRepository(db)

    events, total = repo.list_by_workspace(
        workspace_id=workspace_id,
        page=page,
        page_size=page_size,
        name=name,
        entity_name=entity_name,
        actor=actor,
        timestamp_start=timestamp_start,
        timestamp_end=timestamp_end,
        embedded_site_id=embedded_site_id,
    )

    total_pages = (total + page_size - 1) // page_size if total > 0 else 0

    return ApiResponse.success(
        data=EventListResponse(
            items=[_to_event_response(e) for e in events],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        )
    )


@router.get("/{event_id}", response_model=ApiResponse[EventResponse])
def get_event(
    workspace_code: str,
    event_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EventResponse]:
    """Get a single event by ID."""
    _get_workspace_id(workspace_code, db)
    repo = EventRepository(db)

    event = repo.get_by_id(event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found",
        )

    return ApiResponse.success(data=_to_event_response(event))


@router.post("", response_model=ApiResponse[EventResponse], status_code=status.HTTP_201_CREATED)
def create_event(
    workspace_code: str,
    data: EventCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EventResponse]:
    """Create a new event."""
    workspace_id = _get_workspace_id(workspace_code, db)
    repo = EventRepository(db)

    event = repo.create(workspace_id=workspace_id, user_id=int(current_user.id), data=data)
    return ApiResponse.success(data=_to_event_response(event))


@router.put("/{event_id}", response_model=ApiResponse[EventResponse])
def update_event(
    workspace_code: str,
    event_id: int,
    data: EventUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> ApiResponse[EventResponse]:
    """Update an existing event."""
    _get_workspace_id(workspace_code, db)
    repo = EventRepository(db)

    event = repo.get_by_id(event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found",
        )

    updated_event = repo.update(event, data)
    return ApiResponse.success(data=_to_event_response(updated_event))


@router.delete("/{event_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_event(
    workspace_code: str,
    event_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    """Delete an event (hard delete - permanent removal)."""
    _get_workspace_id(workspace_code, db)
    repo = EventRepository(db)

    event = repo.get_by_id(event_id)
    if not event:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Event not found",
        )

    repo.hard_delete(event)

File: neo/frontend/lib/api/events.ts
/**
 * Events API Client
 * Handles all event related API calls
 */

import type {
	Event,
	CreateEventInput,
	UpdateEventInput,
} from "@/components/event";
import type { ApiResponse } from "@/components/workspace/workspace-types";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || "";

// ============================================================
// API Fetch Wrapper
// ============================================================
async function apiFetch<T>(
	endpoint: string,
	options: RequestInit = {},
): Promise<ApiResponse<T>> {
	const rawToken =
		typeof window !== "undefined" ? localStorage.getItem("neo-auth") : null;
	const parsed = rawToken ? JSON.parse(rawToken) : null;
	const token = parsed?.user?.token ?? parsed?.state?.user?.token ?? null;

	const response = await fetch(`${API_BASE_URL}${endpoint}`, {
		...options,
		headers: {
			"Content-Type": "application/json",
			...(token && { Authorization: `Bearer ${token}` }),
			...options.headers,
		},
		credentials: "include",
	});

	// Handle 204 No Content
	if (response.status === 204) {
		return { code: 0, message: "ok", data: null } as ApiResponse<T>;
	}

	const data = await response.json();

	if (!response.ok) {
		throw {
			code: data.code || response.status,
			message: data.message || "请求失败",
			detail: data.detail,
		};
	}

	if (data.code !== undefined && data.code !== 0) {
		throw {
			code: data.code,
			message: data.message || "请求失败",
			detail: data.detail,
		};
	}

	return data as ApiResponse<T>;
}

// ============================================================
// Events API
// ============================================================

export interface EventListResponse {
	items: Event[];
	total: number;
	page: number;
	page_size: number;
	total_pages: number;
}

export interface ListEventsParams {
	page?: number;
	page_size?: number;
	name?: string;
	entity_name?: string;
	actor?: string;
	timestamp_start?: string;
	timestamp_end?: string;
	embedded_site_id?: number;
}

/**
 * List events for a workspace
 * GET /api/v1/workspaces/{workspace_code}/events
 */
export async function listEvents(
	workspaceCode: string,
	params: ListEventsParams = {},
): Promise<{
	list: Event[];
	total: number;
	page: number;
	page_size: number;
	total_pages: number;
}> {
	const searchParams = new URLSearchParams();
	if (params.page) searchParams.set("page", String(params.page));
	if (params.page_size) searchParams.set("page_size", String(params.page_size));
	if (params.name) searchParams.set("name", params.name);
	if (params.entity_name) searchParams.set("entity_name", params.entity_name);
	if (params.actor) searchParams.set("actor", params.actor);
	if (params.embedded_site_id)
		searchParams.set("embedded_site_id", String(params.embedded_site_id));
	if (params.timestamp_start)
		searchParams.set("timestamp_start", params.timestamp_start);
	if (params.timestamp_end)
		searchParams.set("timestamp_end", params.timestamp_end);

	const query = searchParams.toString();
	const response = await apiFetch<EventListResponse>(
		`/api/v1/workspaces/${workspaceCode}/events${query ? `?${query}` : ""}`,
	);

	return {
		list: response.data?.items || [],
		total: response.data?.total || 0,
		page: response.data?.page || 1,
		page_size: response.data?.page_size || 20,
		total_pages: response.data?.total_pages || 0,
	};
}

/**
 * Get single event
 * GET /api/v1/workspaces/{workspace_code}/events/{id}
 */
export async function getEvent(
	workspaceCode: string,
	eventId: number,
): Promise<Event> {
	const response = await apiFetch<Event>(
		`/api/v1/workspaces/${workspaceCode}/events/${eventId}`,
	);
	return response.data;
}

/**
 * Create event
 * POST /api/v1/workspaces/{workspace_code}/events
 */
export async function createEvent(
	workspaceCode: string,
	data: CreateEventInput,
): Promise<Event> {
	const response = await apiFetch<Event>(
		`/api/v1/workspaces/${workspaceCode}/events`,
		{
			method: "POST",
			body: JSON.stringify(data),
		},
	);
	return response.data;
}

/**
 * Update event
 * PUT /api/v1/workspaces/{workspace_code}/events/{id}
 */
export async function updateEvent(
	workspaceCode: string,
	eventId: number,
	data: UpdateEventInput,
): Promise<Event> {
	const response = await apiFetch<Event>(
		`/api/v1/workspaces/${workspaceCode}/events/${eventId}`,
		{
			method: "PUT",
			body: JSON.stringify(data),
		},
	);
	return response.data;
}

/**
 * Delete event (hard delete)
 * DELETE /api/v1/workspaces/{workspace_code}/events/{id}
 */
export async function deleteEvent(
	workspaceCode: string,
	eventId: number,
): Promise<void> {
	await apiFetch<void>(
		`/api/v1/workspaces/${workspaceCode}/events/${eventId}`,
		{
			method: "DELETE",
		},
	);
}

// ============================================================
// Error Helper
// ============================================================
export function getErrorMessage(code: number): string {
	const messages: Record<number, string> = {
		0: "成功",
		1001: "参数错误",
		1002: "未授权",
		1003: "禁止访问",
		4001: "事件不存在",
		4002: "事件名称已存在",
		3001: "工作区不存在",
		3002: "工作区名称已存在",
		9001: "服务器内部错误",
	};
	return messages[code] || "请求失败，请稍后重试";
}
